File: skills/workspace-knowledge-graph/scripts/discovery.py
#!/usr/bin/env python3

from __future__ import annotations

import os
import re
from pathlib import Path
from typing import Any

from core import IGNORED_DIRS, read_json, read_text, shell, utc_now


def list_git_repos(workspace: Path) -> list[Path]:
    repos: list[Path] = []
    for child in sorted(workspace.iterdir(), key=lambda item: item.name):
        if child.is_dir() and (child / ".git").exists():
            repos.append(child)
    return repos


def count_source_files(root: Path, patterns: dict[str, str]) -> dict[str, int]:
    """单次遍历统计多种源码文件数，避免每个后缀各扫一遍仓库。"""
    counts = dict.fromkeys(patterns, 0)
    for current_root, dirnames, filenames in os.walk(root):
        dirnames[:] = [item for item in dirnames if item not in IGNORED_DIRS]
        for filename in filenames:
            for key, pattern in patterns.items():
                if Path(filename).match(pattern):
                    counts[key] += 1
    return counts


def read_first_heading(path: Path) -> str | None:
    if not path.exists():
        return None
    for line in read_text(path).splitlines():
        stripped = line.strip()
        if stripped.startswith("#"):
            return stripped.lstrip("#").strip()
    return None


def parse_package_json(path: Path) -> dict[str, Any]:
    if not path.exists():
        return {}
    return read_json(path)


def parse_pom_modules(path: Path) -> list[str]:
    if not path.exists():
        return []
    return re.findall(r"<module>([^<]+)</module>", read_text(path))


def list_page_groups(repo_root: Path) -> list[str]:
    pages_dir = repo_root / "src/pages"
    if not pages_dir.exists():
        return []
    groups = [path.name for path in sorted(pages_dir.iterdir()) if path.is_dir()]
    return groups


def list_route_files(repo_root: Path) -> list[str]:
    routes_dir = repo_root / "config/routes"
    if not routes_dir.exists():
        return []
    files = []
    for path in sorted(routes_dir.glob("*.ts")):
        if path.name != "index.ts":
            files.append(str(path.relative_to(repo_root)).replace(os.sep, "/"))
    return files


def list_service_targets(repo_root: Path) -> list[str]:
    services_dir = repo_root / "src/services"
    if not services_dir.exists():
        return []
    targets = [path.name for path in sorted(services_dir.iterdir()) if path.is_dir()]
    return targets


def parse_pnpm_workspace_globs(path: Path) -> list[str]:
    """从 pnpm-workspace.yaml 的 packages 块提取 glob（零依赖的窄解析）。"""
    if not path.exists():
        return []
    globs: list[str] = []
    in_packages = False
    for raw_line in read_text(path).splitlines():
        line = raw_line.split("#", 1)[0].rstrip()
        if not line.strip():
            continue
        stripped = line.strip()
        if not line.startswith((" ", "\t")) and not stripped.startswith("- "):
            in_packages = stripped == "packages:"
            continue
        if in_packages and stripped.startswith("- "):
            value = stripped[2:].strip().strip("'\"")
            if value and not value.startswith("!"):
                globs.append(value)
    return globs


def package_workspace_globs(package_json: dict[str, Any]) -> list[str]:
    """package.json 的 workspaces 字段（yarn / npm workspace，数组或 {packages} 形式）。"""
    workspaces = package_json.get("workspaces")
    if isinstance(workspaces, dict):
        workspaces = workspaces.get("packages")
    if not isinstance(workspaces, list):
        return []
    return [item for item in workspaces if isinstance(item, str) and item and not item.startswith("!")]


def list_workspace_packages(repo_root: Path, package_json: dict[str, Any]) -> list[str]:
    """按声明的 workspace glob 收集包名；无声明时退化为 packages/、apps/ 目录约定。"""
    globs = parse_pnpm_workspace_globs(repo_root / "pnpm-workspace.yaml")
    globs += [item for item in package_workspace_globs(package_json) if item not in globs]
    if not globs:
        globs = ["packages/*", "apps/*"]
    names: list[str] = []
    for pattern in globs:
        if pattern.startswith(("/", "..")):
            continue
        for manifest in sorted(repo_root.glob(f"{pattern.rstrip('/')}/package.json")):
            if "node_modules" in manifest.parts:
                continue
            name = parse_package_json(manifest).get("name")
            if name and name not in names:
                names.append(name)
    return names


def detect_repo_kind(repo_root: Path, package_json: dict[str, Any]) -> str:
    if (repo_root / "pom.xml").exists():
        return "maven-service"
    if (repo_root / "config/config.ts").exists():
        return "bigfish-console"
    if (repo_root / "pnpm-workspace.yaml").exists() or package_workspace_globs(package_json):
        return "node-monorepo"
    # shell 交付类仓库不单设类型：形状因仓库而异，容易过拟合。
    # 该信号由 package_manager=shell-scripts 和 build.sh/release.sh 操作行呈现。
    if (repo_root / "package.json").exists():
        return "node-repo"
    return "unknown"


def shell_scripts_are_primary(package_json: dict[str, Any], repo_root: Path) -> bool:
    if not package_json:
        return False
    shell_entrypoints = [repo_root / "build.sh", repo_root / "release.sh"]
    if not any(path.exists() for path in shell_entrypoints):
        return False
    scripts = package_json.get("scripts", {})
    if not isinstance(scripts, dict):
        return True
    npm_workflow_scripts = {"dev", "start", "build", "test", "lint"}
    return not any(name in scripts for name in npm_workflow_scripts)


def detect_package_manager(package_json: dict[str, Any], repo_root: Path, detected_kind: str) -> str:
    if detected_kind == "maven-service":
        return "maven"
    package_manager = package_json.get("packageManager")
    if isinstance(package_manager, str) and package_manager:
        return package_manager
    if (repo_root / "pnpm-workspace.yaml").exists():
        return "pnpm"
    if not package_json:
        return "N/A"
    if shell_scripts_are_primary(package_json, repo_root):
        return "shell-scripts"
    # 锁文件是明确的包管理器信号；tnpm 只作为没有任何信号时的组织默认。
    for lockfile, manager in (
        ("pnpm-lock.yaml", "pnpm"),
        ("yarn.lock", "yarn"),
        ("bun.lock", "bun"),
        ("bun.lockb", "bun"),
        ("package-lock.json", "npm"),
    ):
        if (repo_root / lockfile).exists():
            return manager
    return "tnpm"


def collect_manifest_paths(repo_root: Path) -> list[str]:
    candidates = [
        repo_root / "package.json",
        repo_root / "pom.xml",
        repo_root / "pnpm-workspace.yaml",
        repo_root / "build.gradle",
        repo_root / "build.gradle.kts",
        repo_root / "go.mod",
        repo_root / "Cargo.toml",
        repo_root / "pyproject.toml",
        repo_root / "README.md",
        repo_root / "config/config.ts",
    ]
    paths = [str(path.relative_to(repo_root)).replace(os.sep, "/") for path in candidates if path.exists()]
    return paths


def find_repo_files(repo_root: Path, filename: str) -> list[Path]:
    """按 IGNORED_DIRS 剪枝查找命名文件，避免扫进 node_modules / target 等目录。"""
    found: list[Path] = []
    for base, dirnames, filenames in os.walk(repo_root):
        dirnames[:] = sorted(item for item in dirnames if item not in IGNORED_DIRS)
        if filename in filenames:
            found.append(Path(base) / filename)
    return found


def sibling_mentions(repo_root: Path, sibling_names: list[str]) -> dict[str, dict[str, Any]]:
    files = find_repo_files(repo_root, "pom.xml")
    if (repo_root / "config/config.ts").exists():
        files.append(repo_root / "config/config.ts")
    mentions: dict[str, dict[str, Any]] = {}
    for sibling in sibling_names:
        total = 0
        matched: list[str] = []
        pattern = re.compile(rf"\b{re.escape(sibling)}\b")
        for path in files:
            try:
                count = len(pattern.findall(read_text(path)))
            except UnicodeDecodeError:
                continue
            if count:
                total += count
                matched.append(str(path.relative_to(repo_root)).replace(os.sep, "/"))
        if total:
            # 把真正命中的文件留作关系证据。
            mentions[sibling] = {"count": total, "files": sorted(matched)}
    return mentions


def build_repo_scan(repo_root: Path, sibling_names: list[str]) -> dict[str, Any]:
    package_json = parse_package_json(repo_root / "package.json")
    readme_title = read_first_heading(repo_root / "README.md")
    detected_kind = detect_repo_kind(repo_root, package_json)
    repo_scan: dict[str, Any] = {
        "name": repo_root.name,
        "path": str(repo_root),
        "remote": shell("git", "remote", "get-url", "origin", cwd=repo_root),
        "detected_kind": detected_kind,
        "package_manager": detect_package_manager(package_json, repo_root, detected_kind),
        "manifests": collect_manifest_paths(repo_root),
        "readme_title": readme_title,
        "counts": count_source_files(
            repo_root,
            {
                "java_files": "*.java",
                "kt_files": "*.kt",
                "ts_files": "*.ts",
                "tsx_files": "*.tsx",
                "py_files": "*.py",
                "go_files": "*.go",
                "rs_files": "*.rs",
            },
        ),
        "package": {
            "name": package_json.get("name"),
            "scripts": sorted(package_json.get("scripts", {}).keys()),
            "script_commands": package_json.get("scripts", {})
            if isinstance(package_json.get("scripts"), dict)
            else {},
        },
        "frontend": {
            "page_groups": list_page_groups(repo_root),
            "route_files": list_route_files(repo_root),
            "service_targets": list_service_targets(repo_root),
        },
        "backend": {
            "modules": parse_pom_modules(repo_root / "pom.xml"),
        },
        # 遗留快照可能还带 "apps" 键；读取方保持 .get 容忍即可。
        "monorepo": {
            "packages": list_workspace_packages(repo_root, package_json),
        },
        "sibling_mentions": sibling_mentions(repo_root, sibling_names),
    }
    return repo_scan


def repo_map(discovery: dict[str, Any]) -> dict[str, dict[str, Any]]:
    return {repo["name"]: repo for repo in discovery["repos"]}


def discover_workspace(workspace: Path) -> dict[str, Any]:
    repos = list_git_repos(workspace)
    sibling_names = [repo.name for repo in repos]
    scanned = [build_repo_scan(repo, [name for name in sibling_names if name != repo.name]) for repo in repos]
    return {
        "generated_at": utc_now(),
        "workspace": {
            "name": workspace.name,
            "path": str(workspace),
            "repo_count": len(repos),
        },
        "repos": scanned,
    }

File: skills/workspace-knowledge-graph/scripts/discovery.ts
#!/usr/bin/env bun

import { existsSync, readFileSync } from "node:fs";
import { basename, join, relative, resolve } from "node:path";

import {
  IGNORED_DIRS,
  isDirectory,
  readJson,
  readText,
  shell,
  sortedDirectory,
  toPosix,
  utcNow,
  walkFiles,
  type JsonObject,
} from "./core.ts";

function isRecord(value: any): value is JsonObject {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function isEmptyRecord(value: JsonObject): boolean {
  return Object.keys(value).length === 0;
}

export function listGitRepos(workspace: string): string[] {
  const repos: string[] = [];
  for (const child of sortedDirectory(workspace)) {
    const childPath = join(workspace, child.name);
    if (isDirectory(childPath) && existsSync(join(childPath, ".git"))) {
      repos.push(childPath);
    }
  }
  return repos;
}

export function countSourceFiles(
  root: string,
  patterns: Record<string, string>,
): Record<string, number> {
  const counts = Object.fromEntries(Object.keys(patterns).map((key) => [key, 0]));
  for (const path of walkFiles(root, { ignoredDirs: IGNORED_DIRS })) {
    const fileName = basename(path);
    for (const [key, pattern] of Object.entries(patterns)) {
      if (pattern.startsWith("*.") && fileName.endsWith(pattern.slice(1))) {
        counts[key] = (counts[key] ?? 0) + 1;
      }
    }
  }
  return counts;
}

export function readFirstHeading(path: string): string | null {
  if (!existsSync(path)) {
    return null;
  }
  for (const line of readText(path).split(/\r?\n/)) {
    const stripped = line.trim();
    if (stripped.startsWith("#")) {
      return stripped.replace(/^#+/, "").trim();
    }
  }
  return null;
}

export function parsePackageJson(path: string): JsonObject {
  return existsSync(path) ? readJson<JsonObject>(path) : {};
}

export function parsePomModules(path: string): string[] {
  if (!existsSync(path)) {
    return [];
  }
  return [...readText(path).matchAll(/<module>([^<]+)<\/module>/g)].map(
    (match) => match[1]!,
  );
}

export function listPageGroups(repoRoot: string): string[] {
  const pagesDir = join(repoRoot, "src/pages");
  if (!existsSync(pagesDir)) {
    return [];
  }
  return sortedDirectory(pagesDir)
    .filter((entry) => isDirectory(join(pagesDir, entry.name)))
    .map((entry) => entry.name);
}

export function listRouteFiles(repoRoot: string): string[] {
  const routesDir = join(repoRoot, "config/routes");
  if (!existsSync(routesDir)) {
    return [];
  }
  return sortedDirectory(routesDir)
    .filter((entry) => entry.name.endsWith(".ts") && entry.name !== "index.ts")
    .map((entry) => toPosix(relative(repoRoot, join(routesDir, entry.name))));
}

export function listServiceTargets(repoRoot: string): string[] {
  const servicesDir = join(repoRoot, "src/services");
  if (!existsSync(servicesDir)) {
    return [];
  }
  return sortedDirectory(servicesDir)
    .filter((entry) => isDirectory(join(servicesDir, entry.name)))
    .map((entry) => entry.name);
}

export function parsePnpmWorkspaceGlobs(path: string): string[] {
  if (!existsSync(path)) {
    return [];
  }
  const globs: string[] = [];
  let inPackages = false;
  for (const rawLine of readText(path).split(/\r?\n/)) {
    const hashIndex = rawLine.indexOf("#");
    const line = (hashIndex >= 0 ? rawLine.slice(0, hashIndex) : rawLine).trimEnd();
    if (!line.trim()) {
      continue;
    }
    const stripped = line.trim();
    if (!line.startsWith(" ") && !line.startsWith("\t") && !stripped.startsWith("- ")) {
      inPackages = stripped === "packages:";
      continue;
    }
    if (inPackages && stripped.startsWith("- ")) {
      const value = stripped
        .slice(2)
        .trim()
        .replace(/^['"]+|['"]+$/g, "");
      if (value && !value.startsWith("!")) {
        globs.push(value);
      }
    }
  }
  return globs;
}

export function packageWorkspaceGlobs(packageJson: JsonObject): string[] {
  let workspaces = packageJson.workspaces;
  if (isRecord(workspaces)) {
    workspaces = workspaces.packages;
  }
  if (!Array.isArray(workspaces)) {
    return [];
  }
  return workspaces.filter(
    (item): item is string =>
      typeof item === "string" && Boolean(item) && !item.startsWith("!"),
  );
}

export function listWorkspacePackages(
  repoRoot: string,
  packageJson: JsonObject,
): string[] {
  const globs = parsePnpmWorkspaceGlobs(join(repoRoot, "pnpm-workspace.yaml"));
  for (const item of packageWorkspaceGlobs(packageJson)) {
    if (!globs.includes(item)) {
      globs.push(item);
    }
  }
  if (globs.length === 0) {
    globs.push("packages/*", "apps/*");
  }
  const names: string[] = [];
  for (const pattern of globs) {
    if (pattern.startsWith("/") || pattern.startsWith("..")) {
      continue;
    }
    const manifestPattern = `${pattern.replace(/\/+$/, "")}/package.json`;
    const manifests = [
      ...new Bun.Glob(manifestPattern).scanSync({
        cwd: repoRoot,
        absolute: true,
        onlyFiles: true,
        followSymlinks: false,
      }),
    ].sort();
    for (const manifest of manifests) {
      if (manifest.split(/[\\/]/).includes("node_modules")) {
        continue;
      }
      const name = parsePackageJson(manifest).name;
      if (name && !names.includes(name)) {
        names.push(name);
      }
    }
  }
  return names;
}

export function detectRepoKind(repoRoot: string, packageJson: JsonObject): string {
  if (existsSync(join(repoRoot, "pom.xml"))) {
    return "maven-service";
  }
  if (existsSync(join(repoRoot, "config/config.ts"))) {
    return "bigfish-console";
  }
  if (
    existsSync(join(repoRoot, "pnpm-workspace.yaml"))
    || packageWorkspaceGlobs(packageJson).length > 0
  ) {
    return "node-monorepo";
  }
  if (existsSync(join(repoRoot, "package.json"))) {
    return "node-repo";
  }
  return "unknown";
}

export function shellScriptsArePrimary(
  packageJson: JsonObject,
  repoRoot: string,
): boolean {
  if (isEmptyRecord(packageJson)) {
    return false;
  }
  if (
    ![join(repoRoot, "build.sh"), join(repoRoot, "release.sh")].some((path) =>
      existsSync(path),
    )
  ) {
    return false;
  }
  const scripts = packageJson.scripts ?? {};
  if (!isRecord(scripts)) {
    return true;
  }
  const npmWorkflowScripts = new Set(["dev", "start", "build", "test", "lint"]);
  return !Object.keys(scripts).some((name) => npmWorkflowScripts.has(name));
}

export function detectPackageManager(
  packageJson: JsonObject,
  repoRoot: string,
  detectedKind: string,
): string {
  if (detectedKind === "maven-service") {
    return "maven";
  }
  const packageManager = packageJson.packageManager;
  if (typeof packageManager === "string" && packageManager) {
    return packageManager;
  }
  if (existsSync(join(repoRoot, "pnpm-workspace.yaml"))) {
    return "pnpm";
  }
  if (isEmptyRecord(packageJson)) {
    return "N/A";
  }
  if (shellScriptsArePrimary(packageJson, repoRoot)) {
    return "shell-scripts";
  }
  for (const [lockfile, manager] of [
    ["pnpm-lock.yaml", "pnpm"],
    ["yarn.lock", "yarn"],
    ["bun.lock", "bun"],
    ["bun.lockb", "bun"],
    ["package-lock.json", "npm"],
  ] as const) {
    if (existsSync(join(repoRoot, lockfile))) {
      return manager;
    }
  }
  return "tnpm";
}

export function collectManifestPaths(repoRoot: string): string[] {
  const candidates = [
    "package.json",
    "pom.xml",
    "pnpm-workspace.yaml",
    "build.gradle",
    "build.gradle.kts",
    "go.mod",
    "Cargo.toml",
    "pyproject.toml",
    "README.md",
    "config/config.ts",
  ];
  return candidates.filter((path) => existsSync(join(repoRoot, path)));
}

export function findRepoFiles(repoRoot: string, filename: string): string[] {
  return [...walkFiles(repoRoot, { ignoredDirs: IGNORED_DIRS })].filter(
    (path) => basename(path) === filename,
  );
}

function readUtf8Strict(path: string): string {
  return new TextDecoder("utf-8", { fatal: true }).decode(readFileSync(path));
}

function isUnicodeWord(character: string | undefined): boolean {
  return character !== undefined && /^[\p{L}\p{N}_]$/u.test(character);
}

function countBoundedLiteral(content: string, literal: string): number {
  const literalCharacters = [...literal];
  const first = literalCharacters[0];
  const last = literalCharacters.at(-1);
  if (first === undefined || last === undefined) {
    return 0;
  }
  let count = 0;
  let offset = 0;
  while (offset <= content.length - literal.length) {
    const index = content.indexOf(literal, offset);
    if (index < 0) {
      break;
    }
    const before = [...content.slice(0, index)].at(-1);
    const after = [...content.slice(index + literal.length)][0];
    if (
      isUnicodeWord(before) !== isUnicodeWord(first)
      && isUnicodeWord(last) !== isUnicodeWord(after)
    ) {
      count += 1;
    }
    offset = index + literal.length;
  }
  return count;
}

export function siblingMentions(
  repoRoot: string,
  siblingNames: string[],
): Record<string, JsonObject> {
  const files = findRepoFiles(repoRoot, "pom.xml");
  if (existsSync(join(repoRoot, "config/config.ts"))) {
    files.push(join(repoRoot, "config/config.ts"));
  }
  const mentions: Record<string, JsonObject> = {};
  for (const sibling of siblingNames) {
    let total = 0;
    const matched: string[] = [];
    for (const path of files) {
      let content: string;
      try {
        content = readUtf8Strict(path);
      } catch {
        continue;
      }
      const count = countBoundedLiteral(content, sibling);
      if (count > 0) {
        total += count;
        matched.push(toPosix(relative(repoRoot, path)));
      }
    }
    if (total > 0) {
      mentions[sibling] = { count: total, files: matched.sort() };
    }
  }
  return mentions;
}

export function buildRepoScan(
  repoRoot: string,
  siblingNames: string[],
): JsonObject {
  const packageJson = parsePackageJson(join(repoRoot, "package.json"));
  const readmeTitle = readFirstHeading(join(repoRoot, "README.md"));
  const detectedKind = detectRepoKind(repoRoot, packageJson);
  const scripts = isRecord(packageJson.scripts) ? packageJson.scripts : {};
  return {
    name: basename(repoRoot),
    path: repoRoot,
    remote: shell("git", "remote", "get-url", "origin", { cwd: repoRoot }),
    detected_kind: detectedKind,
    package_manager: detectPackageManager(packageJson, repoRoot, detectedKind),
    manifests: collectManifestPaths(repoRoot),
    readme_title: readmeTitle,
    counts: countSourceFiles(repoRoot, {
      java_files: "*.java",
      kt_files: "*.kt",
      ts_files: "*.ts",
      tsx_files: "*.tsx",
      py_files: "*.py",
      go_files: "*.go",
      rs_files: "*.rs",
    }),
    package: {
      name: packageJson.name ?? null,
      scripts: Object.keys(scripts).sort(),
      script_commands: scripts,
    },
    frontend: {
      page_groups: listPageGroups(repoRoot),
      route_files: listRouteFiles(repoRoot),
      service_targets: listServiceTargets(repoRoot),
    },
    backend: {
      modules: parsePomModules(join(repoRoot, "pom.xml")),
    },
    monorepo: {
      packages: listWorkspacePackages(repoRoot, packageJson),
    },
    sibling_mentions: siblingMentions(repoRoot, siblingNames),
  };
}

export function repoMap(discovery: JsonObject): Record<string, JsonObject> {
  return Object.fromEntries(
    (discovery.repos ?? []).map((repo: JsonObject) => [repo.name, repo]),
  );
}

export function discoverWorkspace(workspace: string): JsonObject {
  const repos = listGitRepos(workspace);
  const siblingNames = repos.map((repo) => basename(repo));
  const scanned = repos.map((repo) =>
    buildRepoScan(
      repo,
      siblingNames.filter((name) => name !== basename(repo)),
    ),
  );
  return {
    generated_at: utcNow(),
    workspace: {
      name: basename(resolve(workspace)),
      path: resolve(workspace),
      repo_count: repos.length,
    },
    repos: scanned,
  };
}
